### ISSUE TITLE
User-defined geometries and muscle models

### ISSUE THREAD

**Issue by wilhelmbraun** [openworm/sibernetic#165]
Dear all,

I wonder whether it is possible to use Sibernetic with user-generated body geometries and muscle models. I couldn't find any hints on how to do this on this github page.

Thanks a lot for your answer.

Best wishes, Wilhelm

**Comment by skhayrulin** [comment]
@wilhelmbraun Hi and sorry for super long answer. Actually it's quite hard to solve this task in common way. Main reason that depends on complexity final scene geometrically. I mean that final configuration of particles should satisfy many physical conditions. But despite that there was a some try to do that in this [repo](https://github.com/openworm/sibernetic_config_gen) it generate a sibernetic config from [vrml](https://en.wikipedia.org/wiki/VRML) file which you can get from any 3D editor like e.g. blender. In spite this it actually could work with only super simple configuration [example1](https://www.youtube.com/watch?v=fhVlcOVYdzg) [example2](https://www.youtube.com/watch?v=4Q8fOd-7qRs) and as far as I know the work on this didn't moved forward far

**Comment by drojas034** [comment]
@skhayrulin Hello, would that repo work in generating different arrangements for external liquid particles? or how would one be able to generate a custom configuration file without the code crashing?  
Thanks.

**Comment by skhayrulin** [comment]
@drojas034 Hello. As far as I remember it worked fine with  simple configs.

**Comment by drojas034** [comment]
@skhayrulin Yeah I can see it is successful in the examples, great job! (@lgchemer)

Would there be a need to modify it to create something like the image attached? To have a large cube of liquid particles in front of the worm to simulate a worm movement transitioning from no liquid particles into the liquid particles? I was able to move the particles around by modifying a configuration file directly (included below, only erasing initial position and corresponding velocity of liquid particles) with no success. (Could not enqueue read data from buffer  error code is error code is -5)


![Worm 1](https://user-images.githubusercontent.com/55452929/86706136-9450e680-bfe4-11ea-9b07-44f0e533a250.PNG)

[worm_swim_half_resolution_2.zip](https://github.com/openworm/sibernetic/files/4882581/worm_swim_half_resolution_2.zip)

**Comment by clausagerskov** [comment]
@drojas034 It should be possible to write a script to create a scene and then copy the already defined worm into it (looking at the readme for how the scene config should be laid out). I have had success creating simple custom objects from elastic material in blender, exporting to ply files and importing into a scene and adding some water. 
See this folder with an example jupyter notebook: https://www.dropbox.com/sh/px6dsukr5tnrh5u/AADrkYypNeUBz-FJXKRJ-8gBa?dl=0
Creating custom muscle models/body geometries could also be possible in theory but connecting them to the neuron sim is a different story

**Comment by clausagerskov** [comment]
though to create a box of water you would need to contain it with a boundary, which would also block the worm, not sure its possible to create a boundary just blocking water particles.   
Else it should be possible to create a pool that the worm drops into, though it should be noted that sibernetic seems to have segmentation faults at certain large scene sizes/particle numbers

**Comment by clausagerskov** [comment]
@skhayrulin would a selective boundary be feasible to implement?

**Comment by skhayrulin** [comment]
@clausagerskov I'm not sure that I understand correct. You mean to generate configuration from blender file with some complex boundary?

**Comment by clausagerskov** [comment]
nono, just if there are boundary particles that only block water and not the worm, but I just realized that that would not work since the worm also contains water

**Comment by clausagerskov** [comment]
i currently just create boundaries in a jupyter notebook and add in the worm that i copied from the worm config files, the most complex object i have managed to import from blender into sibernetic was a cone

**Comment by skhayrulin** [comment]
boundary particles locate only on boundary plane and do not move through simulation also it blocks every other particles water, elastic. Water inside the worm does not go outside cause other mechanism so-called membranes (triple of adjacent elastic particles).

**Comment by skhayrulin** [comment]
As I admitted before (don't remember when and where :)) It's quite difficult problem to create sibernetic configuration from some 3D model e.g. in blender. Main problem is place the particles on right places. I think man who solve this problem can defend PhD by math, computer science and geometry:)

**Comment by clausagerskov** [comment]
How was the original worm model created?

**Comment by skhayrulin** [comment]
It was hardcoded [here](https://github.com/openworm/sibernetic/blob/WormBodySimulation/src/owHelper.cpp#L95-L1400) by @a-palyanov after I' stored it into file

**Comment by clausagerskov** [comment]
attached is a configuration file i created myself where i throw a cone i imported from blender along the floor into a pool of water. should be possible to isolate the worm from a different config file and put it into this one instead of the cone [cone_pool](https://www.dropbox.com/s/uwnwpwkz44j7zrm/cone_pool?dl=0)

**Comment by clausagerskov** [comment]
created using this notebook: [notebook](https://www.dropbox.com/s/bb3f719vekpf0xm/sibernetic.ipynb?dl=0)
and this file exported from blender: [cone3.ply](https://www.dropbox.com/s/2flb6lafkxlr3k1/cone3.ply?dl=0)
exported from the following blender file:
[cone.blender](https://www.dropbox.com/s/mrhjcweksynjikh/cone.blend?dl=0)

**Comment by pgleeson** [comment]
@clausagerskov The dropbox links you provided in this issue are to files which have been deleted/removed. Do you by any chance have them backed up elsewhere?

**Comment by clausagerskov** [comment]
@pgleeson sure, I just changed provider, here is a link to a folder with those files:
https://e.pcloud.link/publink/show?code=kZjlRTZutqEpmmUaIFQ1ka7mnOP1RSoBJDX
